runtime/compiler/optimizer/HCRGuardAnalysis.cpp

```diff
@@ -88,80 +89,154 @@ bool TR_HCRGuardAnalysis::shouldSkipBlock(TR::Block *block)
-            if (exceptingTTSeen)
-               *(_exceptionGenSetInfo[currentBlock->getNumber()]) |= *(_regularGenSetInfo[currentBlock->getNumber()]);
+            isGen = true;
+            isKill = false;
             }
          }
-      else if (treeTop->getNode()->isTheVirtualGuardForAGuardedInlinedCall() && comp()->cg()->supportsMergingGuards())
+      else if (ttNode->isTheVirtualGuardForAGuardedInlinedCall()
+               && TR_FearPointAnalysis::virtualGuardsKillFear()
+               && comp()->cg()->supportsMergingGuards())
          {
-         TR_VirtualGuard *guardInfo = comp()->findVirtualGuardInfo(treeTop->getNode());
+         TR_VirtualGuard *guardInfo = comp()->findVirtualGuardInfo(ttNode);
          if (guardInfo->getKind() != TR_HCRGuard)
             {
-            _regularKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            if (exceptingTTSeen)
-               _exceptionKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
+            // Theoretically, the guard should only kill its inlined path. However, making it right require adding
+            // complications to the data flow and/or optimizations using the result of the analysis. Based on the
+            // fact that there is few optimization opportunities on the taken side and optimizing it has little benefit,
+            // we require optimizations that can generate fear stay away from taken side such that considering a guard
+            // to be a kill for the taken side is safe.
+            //
+            isKill = true;
+            isGen = false;
             }
          }
+
+      bool canNodeRaiseException = false;
+      if (ttNode->getOpCode().canRaiseException())
+         {
+         canNodeRaiseException = true;
+         }
+      else
+         {
+         // Non-treetop nodes that can raise exceptions
+         if (ttNode->getOpCodeValue() == TR::treetop && ttNode->getNumChildren() > 0)
+            {
+            TR::Node* node = ttNode->getFirstChild();
+            if (!exceptionNodelist.contains(node) && node->getOpCode().canRaiseException())
+               {
+               canNodeRaiseException = true;
+               exceptionNodelist.add(node);
+               }
+            }
+         }
+
+      if (canNodeRaiseException)
+         {
+         // If the exception point is also a yield point, it has to be a gen.
+         // It's possible for the tree to yield, allowing assumptions to be invalidated,
+         // and then throw afterward. An OSR guard (if one is necessary) would only run
+         // after non-exceptional completion of the tree, so it wouldn't stop control from
+         // reaching the exception handler.
+         if (isYieldPoint)
+            hasExceptionGen = true;
+         else
+            {
+            hasExceptionGen = hasExceptionGen || isGen;
+            hasExceptionKill = hasExceptionKill || isKill;
```

Comment:
I think we need the state to be "kill" at _all_ exception points in order to kill along the exception path.

If the first interesting point we encounter in the current block is an exception point, then it's possible to invalidate assumptions, enter the block, and follow an exception edge before there has been a chance to run an OSR guard.

Reply:
Added another bool `hasExceptionNotKill` to fix the issue.